### PR TITLE
feat: add metrics to core-utils 

### PR DESCRIPTION
**Pull request by annieke**
**Description**
We want to be collecting metrics in our TypeScript services. This PR adds a Metrics class to core-utils and uses it in base-service. 

**Additional context**
The metrics class will be used to collect metrics in batch-submitter and data-transport-layer.

### CODE REVIEW

File: packages/core-utils/src/common/metrics.ts
Comment (annieke) [inline packages/core-utils/src/common/metrics.ts:10]: @snario i based `metrics.ts` off of `logger.ts` and [The Graph's Metrics](https://github.com/graphprotocol/common-ts/blob/master/packages/common-ts/src/metrics/index.ts). decided to keep it simple and not include passing in another registry as an option; lmk if you think that might be necessary for us though!

File: packages/core-utils/src/common/metrics.ts
Comment (gakonst) [inline packages/core-utils/src/common/metrics.ts:9]: Should this have a more specific type?

**Comment by gakonst** [comment]
> @gakonst i'll need to use the updated core-utils package in batch-submitter and data-transport-layer, do i need to do anything else besides add the changeset for a new version of core-utils to be published? or can i just use master when it's merged?

No need for you to do anything, once this PR is merged, you'll see that the [`Version Packages`](https://github.com/ethereum-optimism/optimism/pull/494) PR will be updated with the changes from this PR's changeset!

**Comment by gakonst** [comment]
Running this locally, I get:

```
dtl_1              | Well, that's that. We ran into a fatal error. Here's the dump. Goodbye!
dtl_1              | (node:1) UnhandledPromiseRejectionWarning: Error: Invalid metric name
dtl_1              |     at new Metric (/opt/optimism/node_modules/prom-client/lib/metric.js:36:10)
dtl_1              |     at new Counter (/opt/optimism/node_modules/prom-client/lib/counter.js:12:1)
dtl_1              |     at module.exports (/opt/optimism/node_modules/prom-client/lib/metrics/processCpuTotal.js:16:30)
dtl_1              |     at Object.collectDefaultMetrics (/opt/optimism/node_modules/prom-client/lib/defaultMetrics.js:43:3)
dtl_1              |     at new Metrics (/opt/optimism/packages/core-utils/src/common/metrics.ts:29:5)
dtl_1              |     at new BaseService (/opt/optimism/packages/core-utils/src/base-service.ts:29:20)
dtl_1              |     at new L1DataTransportService (/opt/optimism/packages/data-transport-layer/src/services/main/service.ts:47:5)
dtl_1              |     at /opt/optimism/packages/data-transport-layer/src/services/run.ts:25:21
dtl_1              |     at Object.<anonymous> (/opt/optimism/packages/data-transport-layer/src/services/run.ts:63:3)
dtl_1              |     at Module._compile (internal/modules/cjs/loader.js:1063:30)
dtl_1              |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
dtl_1              |     at Module.load (internal/modules/cjs/loader.js:928:32)
dtl_1              |     at Function.Module._load (internal/modules/cjs/loader.js:769:14)
dtl_1              |     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
dtl_1              |     at internal/main/run_main_module.js:17:47
```

File: packages/core-utils/src/common/metrics.ts
Comment (annieke) [reply packages/core-utils/src/common/metrics.ts:9]: prom-client makes it so that it's generic: https://github.com/siimon/prom-client/blame/master/README.md#L83-L91

i'm imagining adding environment as a label but not sure yet, can make an issue to make this type specific as we finalize labels, or remove until we start using labels - what do you think?

**Comment by annieke** [comment]
last commit fixed the dtl build error cc @snario

File: packages/core-utils/src/base-service.ts
Comment (snario) [inline packages/core-utils/src/base-service.ts:29]: Hm, this may be a bit annoying. Does the metrics client dislike spaces in service names? Perhaps we should refactor the `BaseService` such that names cannot have spaces in them so that the logs and metrics use identical names.

Comment (annieke) [reply packages/core-utils/src/base-service.ts:29]: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels only underscores and colons 😬 honestly might be the move for cleanliness sake - thoughts on importing this validation and enforcing it? https://github.com/siimon/prom-client/blob/96f7495d66b1a21755f745b1367d3e530668a957/lib/validation.js#L5-L11 or should our validation be looser?

Comment (annieke) [reply packages/core-utils/src/base-service.ts]: just renamed everything without adding the validation since we will get the error on metrics initialization

**Comment by snario** [comment]
Latest `integration` job is [failing](https://github.com/ethereum-optimism/optimism/pull/503/checks?check_run_id=2396180382), seems to be the #430?

**Comment by gakonst** [comment]
The failure was due to not renaming message relayer metric, which resulted in the message relayer not being brought up. Fixed in https://github.com/ethereum-optimism/optimism/pull/503/commits/c15a77a2624eb3f45a7026d33bdf0eb5ed4c0349

**Review by gakonst — APPROVED** [review]
LGTM

**Comment by maurelian** [comment]
I've removed myself as a core-utils CODEOWNER for the time being.

**Pull request merged**
